File: digital/avr/modules/proto/proto.c
/* proto.c */
/* avr.proto - Protocol AVR module. {{{
 *
 * Copyright (C) 2005 Nicolas Schodet
 *
 * Robot APB Team/Efrei 2006.
 *        Web: http://assos.efrei.fr/robot/
 *      Email: robot AT efrei DOT fr
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 * }}} */
#include "common.h"
#include "proto.h"

#include <ctype.h>

/* +AutoDec */

/** Accept a digit to be used for args. */
static void
proto_accept_digit (uint8_t c);

#if AC_PROTO_QUOTE == 1
/** Accept a quoted char to be used for args. */
static void
proto_accept_char (uint8_t c);
#endif

/* Send a hex digit. */
inline static void
proto_hex (uint8_t h);

/* -AutoDec */

static uint8_t cmd;
static uint8_t size;
static uint8_t args[AC_PROTO_ARGS_MAX_SIZE];

/** Step of decoding:
 *  - 0: nothing received.
 *  - 1: bang received.
 *  - 2: command received.
 *  - 3: command received, and processing a number.
 *  - 4: quote received. 
 *  - 5: double quote received. */
static uint8_t step;

/** Accept a new character. */
void
proto_accept (uint8_t c)
{
    if (c == '!')
	step = 1;
    else
      {
	switch (step)
	  {
	  case 0:
	    /* Nothing received yet. */
	    break;
	  case 1:
	    /* Bang received yet. */
	    if (isalpha (c))
	      {
		cmd = c;
		size = 0;
		step = 2;
	      }
	    else
	      {
		AC_PROTO_CALLBACK ('?', 0, 0);
		step = 0;
	      }
	    break;
	  case 2:
	    /* Command received yet. */
	    if (c == '\r')
	      {
		AC_PROTO_CALLBACK (cmd, size, args);
		step = 0;
	      }
#if AC_PROTO_QUOTE == 1
	    else if (c == '\'')
		step = 4;
	    else if (c == '"')
		step = 5;
#endif
	    else
	      {
		step = 3;
		proto_accept_digit (c);
	      }
	    break;
	  case 3:
	    step--;
	    proto_accept_digit (c);
	    break;
#if AC_PROTO_QUOTE == 1
	  case 4:
	    step = 2;
	    proto_accept_char (c);
	    break;
	  case 5:
	    if (c == '\r')
	      {
		AC_PROTO_CALLBACK (cmd, size, args);
		step = 0;
	      }
	    else
	      {
		proto_accept_char (c);
	      }
	    break;
#endif
	  }
      }
}

/** Accept a digit to be used for args. */
static void
proto_accept_digit (uint8_t c)
{
    /* Test for argument list overflow. */
    if (size >= AC_PROTO_ARGS_MAX_SIZE)
      {
	AC_PROTO_CALLBACK ('?', 0, 0);
	step = 0;
	return;
      }
    /* Convert from hexa. */
    if ('0' <= c && c <= '9')
	c -= '0';
    else if ('a' <= c && c <= 'f')
	c -= 'a' - 10;
    else if ('A' <= c && c <= 'F')
	c -= 'A' - 10;
    else
      {
	AC_PROTO_CALLBACK ('?', 0, 0);
	step = 0;
	return;
      }
    /* Add digit. */
    args[size] <<= 4;
    args[size] |= c;
    if (step == 2)
	size++;
}

#if AC_PROTO_QUOTE == 1
/** Accept a quoted char to be used for args. */
static void
proto_accept_char (uint8_t c)
{
    /* Test for argument list overflow or unwanted char. */
    if (size >= AC_PROTO_ARGS_MAX_SIZE || !isprint (c))
      {
	AC_PROTO_CALLBACK ('?', 0, 0);
	step = 0;
	return;
      }
    /* Add char. */
    args[size] = c;
    size++;
}
#endif

/* Send a hex digit. */
inline static void
proto_hex (uint8_t h)
{
    AC_PROTO_PUTC (h >= 10 ? h - 10 + 'a' : h + '0');
}

/* Send a argument byte. */
void
proto_arg (uint8_t a)
{
    proto_hex ((a >> 4) & 0xf);
    proto_hex ((a >> 0) & 0xf);
}

/** Send a command, generic function. */
void
proto_send (uint8_t cmd, uint8_t size, uint8_t *args)
{
    AC_PROTO_PUTC ('!');
    AC_PROTO_PUTC (cmd);
    while (size--)
	proto_arg (*args++);
    AC_PROTO_PUTC ('\r');
}

File: host/proto/proto.py
# proto - Proto interface. {{{
#
# Copyright (C) 2008 Nicolas Schodet
#
# APBTeam:
#        Web: http://apbteam.org/
#      Email: team AT apbteam DOT org
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
# }}}
"""Proto interface module."""
import binascii, struct, select

START = 0
BANG = 1
CMD = 2
ARG = 3

class Proto:

    def __init__ (self, file, date, timeout, log = None):
	"""Initialise and set file (serial port, pty, socket...), date
	function and timeout value.

	- file: open file connected to the slave device.
	- date: when called, should return the current time.
	- timeout: time after which retransmission is done.
	- log: if defined, will be called with a log string.

	"""
	self.file = file
	self.date = date
	self.last_send = None
	self.timeout = timeout
	self.send_queue = [ ]
	self.state = START
	self.log = log
	self.handlers = { }

    def send (self, *frame):
	"""Queue a frame to send."""
	if not self.send_queue:
	    self.last_send = None
	self.send_queue.append (Frame (*frame))

    def read (self):
	"""Read from file and receive frames."""
	for f in self.recv ():
	    if self.log:
		self.log ('recv %s' % f)
	    if self.send_queue and f == self.send_queue[0]:
		del self.send_queue[0]
		if self.send_queue:
		    self.send_head ()
	    else:
		self.dispatch (f)

    def sync (self):
	"""Send frames, return True if all is sent."""
	if self.send_queue and (self.last_send is None
		or self.last_send + self.timeout < self.date ()):
	    self.send_head ()
	return not self.send_queue

    def wait (self, cond = None):
	"""Wait forever or until cond () is True."""
	while not (self.sync () and (cond is not None or cond ())):
	    fds = select.select ((self,), (), (), self.timeout)[0]
	    for i in fds:
		assert i is self
		i.read ()

    def register (self, command, fmt, handler):
	"""Register a handler for the specified command and format.  The
	handler will receive decoded arguments."""
	key = (command, struct.calcsize ('!' + fmt))
	assert key not in self.handlers
	self.handlers[key] = (handler, fmt)

    def fileno (self):
	"""Return file descriptor, for use with select."""
	return self.file.fileno ()

    def send_head (self):
	"""Send first frame from the send queue."""
	if self.log:
	    self.log ('send %s' % self.send_queue[0])
	self.file.write (self.send_queue[0].data ())
	self.last_send = self.date ()

    def recv (self):
	"""Receive a frame, used as a generator."""
	for c in self.file.read (1):
	    if c == '!':
		self.state = BANG
	    else:
		if self.state == START:
		    pass
		elif self.state == BANG:
		    if c.isalpha ():
			self.recv_command = c
			self.recv_args = ''
			self.state = CMD
		    else:
			self.recv_error ()
		elif self.state == CMD:
		    if c == '\r':
			f = Frame (self.recv_command)
			f.args = binascii.unhexlify (self.recv_args)
			yield f
		    elif (c >= '0' and c <= '9') or (c >= 'a' and c <= 'f'):
			self.recv_args += c
			self.state = ARG
		    else:
			self.recv_error ()
		else:
		    assert self.state == ARG
		    if (c >= '0' and c <= '9') or (c >= 'a' and c <= 'f'):
			self.recv_args += c
			self.state = CMD
		    else:
			self.recv_error ()

    def recv_error (self):
	"""Handle reception errors."""
	self.state = START
	if self.log:
	    self.log ('error')
	# Resend now.
	if self.send_queue:
	    self.send_head ()

    def dispatch (self, frame):
	"""Pass a received frame to the correct handler."""
	key = (frame.command, len (frame.args))
	if key in self.handlers:
	    h = self.handlers[key]
	    h[0] (*(frame.decode (h[1])))

class Frame:

    def __init__ (self, command = None, fmt = '', *args):
	"""Initiliase a frame.  If command is given, the frame is constructed
	using a struct.pack like fmt string."""
	if command:
	    assert len (command) == 1 and command.isalpha ()
	    self.command = command
	    self.args = struct.pack ('!' + fmt, *args)
	else:
	    self.command = None
	    self.args = ''
    
    def data (self):
	"""Get a frame representation ready to send."""
	return '!' + self.command + binascii.hexlify (self.args) + '\r'

    def decode (self, fmt):
	"""Decode using a struct.unpack like fmt string."""
	return struct.unpack ('!' + fmt, self.args)

    def __eq__ (self, other):
	"""Compare for equality."""
	return self.command == other.command and self.args == other.args

    def __str__ (self):
	"""Convert to string."""
	return '!' + self.command + binascii.hexlify (self.args)
